crates/sui-framework/packages/sui-framework/sources/config.move

```diff
@@ -0,0 +1,229 @@
+    #[allow(unused_const)]
+    const EBCSSerializationFailure: u64 = 2;
+
+    public struct Config<phantom WriteCap> has key {
+        id: UID,
+    }
+
+    public struct Setting<Value: copy + drop + store> has store {
+        data: Option<SettingData<Value>>,
+    }
+
+    public struct SettingData<Value: copy + drop + store> has store {
+        newer_value_epoch: u64,
+        newer_value: Value,
+        older_value_opt: Option<Value>,
+    }
+
+    public(package) fun new<WriteCap>(_cap: &mut WriteCap, ctx: &mut TxContext): Config<WriteCap> {
+        Config<WriteCap> { id: object::new(ctx) }
+    }
+
+    #[allow(lint(share_owned))]
+    public(package) fun create<WriteCap>(cap: &mut WriteCap, ctx: &mut TxContext) {
+        transfer::share_object(new<WriteCap>(cap, ctx))
+    }
+
+    public(package) fun new_for_epoch<
+        WriteCap,
+        Name: copy + drop + store,
+        Value: copy + drop + store,
+    >(
+        config: &mut Config<WriteCap>,
+        _cap: &mut WriteCap,
+        name: Name,
+        value: Value,
+        _ctx: &mut TxContext,
+    ): Option<Value> {
+        let epoch = _ctx.epoch();
+        if (!field::exists_(&config.id, name)) {
+            let sobj = Setting {
+                data: option::some(SettingData {
+                    newer_value_epoch: epoch,
+                    newer_value: value,
+                    older_value_opt: option::none(),
+                }),
+            };
+            field::add(&mut config.id, name, sobj);
+            option::none()
+        } else {
+            let sobj: &mut Setting<Value> = field::borrow_mut(&mut config.id, name);
+            let SettingData {
+                newer_value_epoch,
+                newer_value,
+                older_value_opt,
+            } = sobj.data.extract();
+            assert!(epoch > newer_value_epoch, EAlreadySetForEpoch);
+            sobj.data.fill(SettingData {
+                newer_value_epoch: epoch,
+                newer_value: value,
+                older_value_opt: option::some(newer_value),
+            });
+            older_value_opt
+        }
+    }
+
+    public(package) fun exists_with_type<
+        WriteCap,
+        Name: copy + drop + store,
+        Value: copy + drop + store,
+    >(
+        config: &Config<WriteCap>,
+        name: Name,
+    ): bool {
+        field::exists_with_type<_, Setting<Value>>(&config.id, name)
+    }
+
+    #[allow(unused_mut_parameter)]
+    public(package) fun exists_with_type_for_epoch<
+        WriteCap,
+        Name: copy + drop + store,
+        Value: copy + drop + store,
+    >(
+        config: & Config<WriteCap>,
+        name: Name,
+        ctx: &TxContext,
+    ): bool {
+        field::exists_with_type<_, Setting<Value>>(&config.id, name) && {
+            let epoch = ctx.epoch();
+            let sobj: &Setting<Value> = field::borrow(&config.id, name);
+            epoch == sobj.data.borrow().newer_value_epoch
```

Comment:
Was this comment for `new_for_epoch`? 
If so, I wanted to give something that indicated it would abort if already set, hence "new". Maybe `add`? Not sure

Reply:
Not it was for this function -- I think the thing I got confused about when I first read this was whether it could ever be in an "unset" state in which case this function's name has a different meaning (is there a current value assigned to this `name` in the current epoch). But I realize now there must always be a value, so the name isn't as bad. 

That being said though, I still feel as though the name here is a bit hard to follow. IMO `set_in_current_epoch_for_type` or something like that would be a bit easier to follow.